File: binary-search-trees/tree.js
var Tree = function(){
  this.root = null;
}

var Node = function(val){
  this.val = val;
  this.left = null;
  this.right = null;
}

Tree.prototype.bfs = function(){
  var queue = [this.root];
  var results = [];

  while(queue.length){
    var current = queue.shift();
    results.push(current.val);
    if(current.left) queue.push(current.left);
    if(current.right) queue.push(current.right);
  }

  return results;
}

Tree.prototype.dfsPreOrder = function(node, results){
  results.push(node.val);
  if(node.left) this.dfsPreOrder(node.left, results);
  if(node.right) this.dfsPreOrder(node.right, results);

  return results;
}

var tree = new Tree();
tree.root = new Node(8);
tree.root.left = new Node(3);
tree.root.right = new Node(10);
tree.root.left.left = new Node(1);
tree.root.left.right = new Node(4);

// console.log(tree.dfsPreOrder(tree.root, []));

// Tree.prototype.dfsPre = function(node, arr, search){
//   if(!node) return;
//   // console.log(node.val, search);
//   if(node.val === search) return node;
//
//   var result = arr;
//   result.push(node.val);
//   this.dfsPre(node.left, result, search);
//   this.dfsPre(node.right, result, search);
//
//   return result;
// }
Tree.prototype.dfs = function(node, search){
  if(!node) return;
  if(node.val === search) return node;

  if(node.val > search && node.left){
    return this.dfs(node.left, search);
  }
  if(node.val < search && node.right){
    return this.dfs(node.right, search);
  }
  return "Not in tree";
}

console.log(tree.dfs(tree.root, 11));


//
// Tree.prototype.dfsIn = function(node, arr, search){
//   if(!node) return;
//   // console.log(node.val, search);
//   // if(node.val === search) return node;
//
//   var result = arr;
//   this.dfsPre(node.left, result, search);
//   result.push(node.val);
//   this.dfsPre(node.right, result, search);
//
//   return result;
// }
//
// Tree.prototype.dfsPost = function(node, arr, search){
//   if(!node) return;
//   // console.log(node.val, search);
//   // if(node.val === search) return node;
//
//   var result = arr;
//   this.dfsPre(node.left, result, search);
//   this.dfsPre(node.right, result, search);
//   result.push(node.val);
//
//   return result;
// }
//
//
//
// // console.log(tree.bfs());
// console.log(tree.dfsPre(tree.root, 4));
// // console.log(tree.dfsIn(tree.root, [], 4));
// // console.log(tree.dfsPost(tree.root, [], 4));

File: binary-search-trees/Tree.java
import java.util.*;

public class Tree{
  public static void main(String[] args){

    Tree binTree = new Tree();

    binTree.insertIteratively(5);
    binTree.insertIteratively(3);
    binTree.insertIteratively(4);
    // System.out.println(binTree.root.value);
    // System.out.println(binTree.root.right.value);
    // System.out.println(binTree.root.left);
    binTree.insertIteratively(2);
    binTree.insertIteratively(7);
    binTree.insertIteratively(6);
    // System.out.println(binTree.root.left.value);
    // System.out.println(binTree.containsIteratively(2));
    // System.out.println(binTree.containsIteratively(5));
    // System.out.println(binTree.containsIteratively(1));
    // System.out.println(binTree.containsIteratively(3));
    // System.out.println(binTree.containsIteratively(4));
    System.out.println(binTree.breadthFirstSearch());
    System.out.println(binTree.dfsPre());

  }

  Node root = null;

  public Tree(){}

  public void insertIteratively(int val){
    Node newNode = new Node(val, null, null);

    if(this.root == null){
      this.root = newNode;
    } else {
      Node current = this.root;

      while(true){
        if(val < current.value) {
          if(current.left == null){
            current.left = newNode;
            break;
          } else {
            current = current.left;
          }
        } else if(val > current.value){
          if(current.right == null){
            current.right = newNode;
            break;
          } else {
            current = current.right;
          }
        } else if(val == current.value){
          System.out.println("Duplicate");
          break;
        }
      }
    }
  }

  public void insertRecursively(int val){
    Node newNode = new Node(val, null, null);

    if(this.root == null) this.root = newNode;


  }

  public boolean containsIteratively(int val){
    Node current = this.root;

    while(current != null){
      if(current.value == val) return true;
      else if(val < current.value) current = current.left;
      else if(val > current.value) current = current.right;
    }

    return false;
  }

  public ArrayList<Integer> breadthFirstSearch(){
    Node node = this.root;
    ArrayList<Node> queue = new ArrayList<Node>();
    queue.add(node);
    ArrayList<Integer> data = new ArrayList<Integer>();

    while(queue.size() > 0){
      node = queue.get(0);
      data.add(node.value);
      if(node.left != null) queue.add(node.left);
      if(node.right != null) queue.add(node.right);
      queue.remove(0);
    }

    return data;
  }

  public ArrayList<Integer> dfsPre(){
    ArrayList<Integer> data = new ArrayList<Integer>();
    Node current = this.root;

    search(current, data);

    return data;
  }

  public static void search(Node n, ArrayList<Integer> data){
    data.add(n.value);
    if(n.left != null) search(n.left, data);
    if(n.right != null) search(n.right, data);
  }


}
